patient.go

```diff
@@ -0,0 +1,107 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	patient := &model.Patient{}
+	resp, err := p.client.Do(ctx, req, patient)
+
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return patient, resp, nil
+}
+
+// PatientSearchOptions is the options we pass into the request for searching a patient
+type PatientSearchOptions struct {
+	// A fuzzy search is performed, including checks for homophones, transposed names and historic information.
+	// You cant use wildcards with fuzzy search
+	FuzzyMatch *bool `url:"_fuzzy-match,omitempty"`
+	// The search only returns results where the score field is 1.0. Use this with care - it is unlikely to work with fuzzy search or wildcards.
+	ExactMatch *bool `url:"_exact-match,omitempty"`
+	// The search looks for matches in historic information such as previous names and addresses.
+	// This parameter has no effect for a fuzzy search, which always includes historic information.
+	History *bool `url:"_history,omitempty"`
+	// For application-restricted access, this must be 1
+	MaxResults int `url:"_max-results"`
+	// if used with wildcards, fuzzy match must be false. Wildcards must contain at least two characters, this matches Smith, Smythe. Not case-sensitive.
+	Family *string `url:"family,omitempty"`
+	// The patients given name, can be used with wildcards. E.g. Jane Anne Smith
+	// Use * as a wildcard but not in the first two characters and not in fuzzy search mode
+	Given  *[]string `url:"given,omitempty"`
+	Gender *Gender   `url:"gender,omitempty"`
+	// Format: <eq|ge|le>yyyy-mm-dd e.g. eq2021-08-01
+	BirthDate []*string `url:"birthdate,omitempty"`
+	// For a fuzzy search, this is ignored for matching but included in the score calculation.
+	// Format: <eq|ge|le>yyyy-mm-dd e.g. eq2021-08-01
+	DeathDate *[]string `url:"death-date,omitempty"`
+	// Not case sensitive. Spaces are ignored, for example LS16AE and LS1 6AE both match LS1 6AE
+	Postcode *string `url:"address-postcode,omitempty"`
+	// The Organisation Data Service (ODS) code of the patient's registered GP practice.
+	// Not case sensitive. For a fuzzy search, this is ignored for matching but included in the score calculation.
+	// Example: Y12345
+	GeneralPractioner *string `url:"general-practitioner,omitempty"`
+}
+
+// Search searches for a patient in the PDS
+// The behaviour of this endpoint depends on your access mode:
+//https://digital.nhs.uk/developer/api-catalogue/personal-demographics-service-fhir#api-Default-search-patient
+func (p *PatientService) Search(ctx context.Context, opts PatientSearchOptions) ([]*model.Patient, *Response, error) {
+	url, err := addParamsToUrl(path, opts)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req, err := p.client.NewRequest(http.MethodGet, url, nil)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	result := &model.Result{}
+
+	resp, err := p.client.Do(ctx, req, result)
+
+	if err != nil {
+		return nil, resp, err
+	}
+
+	if len(result.Entry) == 0 {
+		return nil, resp, errors.New("user not found")
```

Comment:
Don't error here. We need to be able to distinguish between a "not found" and an actual error

Reply:
I agree here. I'm thinking about expanding more on the errors. Specifically about creating an APIResponseError which would contain the response body inside the error. That way users will know specifically what is wrong either with their specific call or the underlying API.

Reply:
Another future error could be RateLimitError, etc.